buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java

```diff
@@ -0,0 +1,331 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.gradle;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.unmodifiableList;
+
+/**
+ * Parse the Java source file containing the versions declarations and use the known rules to figure out which are all
+ * the version the current one is wire and index compatible with.
+ * On top of this, figure out which of these are unreleased and provide the branch they can be built from.
+ * <p>
+ * Note that in this context, currentVersion is the unreleased version this build operates on.
+ * At any point in time there will surely be four such unreleased versions being worked on,
+ * thus currentVersion will be one of these:
+ * - the unreleased <b>major</b>, a+1.0.0 on the `master` branch
```

Comment:
I think I see what you meant to do with `a`. Is this supposed to be the currently released major version? If so, I would call it `M` (and define it above). This way it can be easily distinguished below with `M.x` where a capitalized variable is a constant that would be substituted (otherwise there is no way to understand x is not to be substituted as well).

Reply:
That's right,  good point, changing to `M`.